toolkit/tools/pkg/imagecustomizerlib/extractpartitions.go

```diff
@@ -0,0 +1,87 @@
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagegen/diskutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/file"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+// Extract all partitions of connected image into separate files with specified format.
+func extractPartitions(imageConnection *ImageConnection, outputImageFile string, partitionFormat string) error {
+
+	// Extract basename from outputImageFile. E.g. if outputImageFile is "image.qcow2", then basename is "image".
+	basename := strings.TrimSuffix(filepath.Base(outputImageFile), filepath.Ext(outputImageFile))
+
+	// Get path of loop device associated with the image.
+	imageLoopDevice := imageConnection.Loopback().DevicePath()
+
+	// Get output directory path.
+	outDir := filepath.Dir(outputImageFile)
+
+	var err error
+
+	// Get partition info.
+	diskPartitions, err := diskutils.GetDiskPartitions(imageLoopDevice)
+	if err != nil {
+		return err
+	}
+
+	// Start extracting from 1 because the diskPartitions[0] refers to the image itself.
+	for partitionNum := 1; partitionNum < len(diskPartitions); partitionNum++ {
+		rawFilename := basename + "_" + strconv.Itoa(partitionNum) + ".raw"
+		partitionLoopDevice := diskPartitions[partitionNum].Path
+
+		err, partitionFilepath := createRawFile(outDir, partitionLoopDevice, rawFilename)
+		if err != nil {
+			return err
+		}
+
+		if partitionFormat == "raw-zstd" {
+			err, partitionFilepath = compressWithZstd(partitionFilepath)
+			if err != nil {
+				return err
+			}
+		}
+
+		logger.Log.Infof("Partition file created: %s", partitionFilepath)
+	}
+	return nil
+}
+
+// Creates .raw file for the mentioned partition path.
+func createRawFile(outDir, devicePath, name string) (err error, filename string) {
+	const (
+		defaultBlockSize = 1024 * 1024 // 1MB
+		squashErrors     = true
+	)
+
+	fullPath := filepath.Join(outDir, name)
+	ddArgs := []string{
+		fmt.Sprintf("if=%s", devicePath),       // Input file.
+		fmt.Sprintf("of=%s", fullPath),         // Output file.
+		fmt.Sprintf("bs=%d", defaultBlockSize), // Size of one copied block.
+	}
+
+	return shell.ExecuteLive(squashErrors, "dd", ddArgs...), fullPath
```

Comment:
You should add error messages as the error flows up the stack so that if the tool fails, you know exactly where it failed.

For example,

```golang
err := shell.ExecuteLive(squashErrors, "dd", ddArgs...)
if err != nil {
    return "", fmt.Errorf("failed to copy block device into file:\n%w", err)
}

return fullPath, nil
```

You should do the same in other places as well.

For most `if err != nil` checks, you should add extra information to the error. Though, try to avoid adding duplication information.

Reply:
great feedback, appreciate it!